### DISCUSSION TITLE
Using sensors to implement complex operations

### DISCUSSION THREAD

**Discussion by UqamProvencherRobert** [StackStorm/st2#6175] [General]
_If the current topic is not suited for this forum, please write me so and I'll remove it._

We are considering using St2 to automate something like a hundred workflows.
Most of these wf would be triggered by an http request telling us something happened in an application.

But with the basic understanding I have of St2, it seems that an http request would be caught by a sensor.
Like the echo_flask_app sensor in the Examples pack.
The answer to the request is sent by the sensor. Not the following action(s) defined by a rule.

So if the answers given to the requests need informations from the operations we implement,
we would have to code all our things in sensors and not use trigers-rules-workflows-actions.

The most basic example is if something goes wrong during the execution of an operation, we would want to answer a
500 Internal Server Error to the request. How to do that if our code is in actions ?

So, finally, I'd like to ask these questions :

In the St2 paradigm/philosophy is it ok to code everything in sensors, even complex operations, not using rules-actions... ?
Does that defeat the purpose of using St2 ?

Is there a way to use the whole sensor-trigger-rule-workflow-actions sequence to receive an http request with a sensor and at the last action answer this request ?

Thanks for your time!

**Comment by fdrab** [comment]
you can also call the /executions api and get an execution ID, no?

**Reply by UqamProvencherRobert** [reply]
I'm not sure I understand your meaning.
Like, in the sensor, wait for the actions to complete and then check their status, using the apis ?
Interresting! But I don't know if such logic would go agaist the st2 paradigm...

**Reply by fdrab** [reply]
not sure I understand your post correctly. You have an application that will send a HTTP request in case a specific event happens. In ST2 there are at least 2 ways to execute content:
1, webhooks, where you call a custom endpoint that will then be evaluated by rules and if the criteria match, will launch a flow / action execution
2, directly calling /executions API and bypass the rules completely by executing the specific workflow / action by passing inputs and what to execute directly

The webhooks method currently does not provide any response besides that the call was accepted. The execution API will give you a JSON object containing details of the execution which you can later query to see what the result was.

**Reply by UqamProvencherRobert** [reply]
Thanks for trying to understand my questions.

I'm talking about what you wrote 
     "The webhooks method currently does not provide any response besides that the call was accepted"

Is there a way to get around that while respecting the st2 paradigm?

Like the use case you wrote, lets say an app sends a http req, st2 receives it and launch a workflow (by rule or exeution api).
What if the application that sent the http request needs to know if the workflow was successful?

It seems to me that it would involve loads of code in the sensor. 
The sensor would have to wait for all the actions in the wf to complete before sending the http response to the application.

Hope it's a bit more clear...

**Reply by fdrab** [reply]
The app should use the execution api. Execution api returns an execution ID that the APP then can query. There's no way currently to get a result of execution launched via webhooks, as webhooks execution are not guaranteed (rule may not match). Furthermore, I find it a bit strange that the app would send a HTTP request and wait for a response from the server until the execution finishes. The app can't know how long the execution will run and would "hang" on the http waiting for the response, and not know whether the flow is still running, or there was a time-out.

**Reply by UqamProvencherRobert** [reply]
Well, thanks for answering. Good point, about the application hanging. I'll think about all this...

**Reply by dalesmith** [reply]
@UqamProvencherRobert, I'd be happy to help you explore what is possible with ST2 and how best to accomplish your intended outcomes. *** Fair Disclosure *** My affiliation is with a company (orchestral.ai) that provides a commercially supported version of ST2 called - "Composer". We work with enterprise brands to deploy automation and orchestration capabilities across a broad range of use cases. Happy to explore further with you if interested.

**Reply by UqamProvencherRobert** [reply]
Thank you for the offer @dalesmith :)
